### ISSUE TITLE
Slow response when adding multiple (thousands) files.

### ISSUE THREAD

**Issue by Vogelhaus** [PhileCMS/Phile#256]
Hi,

I am evaluating the performance of PhileCMS. Overall a great "simple" CMS. Unfortunately I encountered   the problem that if you add a lot of content pages ( I am talking about 100 / 1000) it slows all down. Requests take 10s for 10 000 pages.

Why is it necessary for PhileCMS to loop over all the content pages on each request?

**Comment by NeoBlack** [comment]
I agree, that we can maybe improve the handling, but the idea for PhileCMS was never to have big sites.
It was designed for small websites. PhileCMS works without any kind of database, if you have any suggestion how to improve, I would like hear your ideas.

**Comment by kicken** [comment]
> Why is it necessary for PhileCMS to loop over all the content pages on each request?

I believe the culprit is the `{{pages}}` variable for the template.  In order to create it Phile needs to do a recursive search for all available content pages.

If you do not need this variable (ie, navigation is handled some other way) then disabling it should speed things up.

Perhaps a setting to skip populating this variable, or some way to make it only populate if used would be beneficial?

**Comment by james2doyle** [comment]
Can we cache the `{{ pages }}` call also?

**Comment by Vogelhaus** [comment]
It would be also helpful if $pages was more specific. Maybe I just want the pages of a specific content folder. like $pages["documentation"]

**Comment by Schlaefer** [comment]
> It would be also helpful if $pages was more specific. Maybe I just want the pages of a specific content folder. like $pages["documentation"]

You can test for page meta-data if you only want to act on a subset. E.g. with twig:

```twig
{% for page in pages %}
    {% if page.folder == 'documentation' %}
    	// do if documentation page …
```

This default behavior should serve most people well out of the box, but depends on pages `knowing` about all pages. If the default pages variable doesn't fit your needs its easy to create your own in a custom plugin. Use the `findAll` method you already found to grab what you want and populate a template variable with it.

**Comment by Vogelhaus** [comment]
> This default behavior should serve most people well out of the box, but depends on pages knowing about all pages. If the default pages variable doesn't fit your needs its easy to create your own in a custom plugin. Use the findAll method you already found to grab what you want and populate a template variable with it.

And this is I think the reason why it could be inefficient for some use cases. I don't say its totally "bad" and I am sure it suits smaller web pages very well. But in case you have a deep directory structure it will have to a lot of I/O cycles to retrieve all the pages in these folders.

So maybe even smaller sites with around 100 pages in 25 different folders could be potentially affected. I will try that out.

**Comment by Schlaefer** [comment]
I benchmarked recently¹ and up to 100 pages should be fine. After that it becomes a question of server, traffic, cache engine etc. A few hundred pages may be doable. My educated guess is that at 1000+ pages code tweaking is always required.

¹For this [this cache plugin](https://github.com/Schlaefer/phileTotalCache). If you're performance sensitive, you could give it a try.

**Comment by Schlaefer** [comment]
fixed in #257
